apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+    std::vector<VarOrRVar> all_vars;
+    all_vars.reserve(serial_rvars.size() + 4);
+    for (RVar v : serial_rvars) {
+        all_vars.push_back(v);
+    }
+    if (!r_gpu_threads.name().empty()) {
+        all_vars.push_back(r_gpu_threads);
+    }
+    if (!gpu_threads.name().empty()) {
+        all_vars.push_back(gpu_threads);
+    }
+    if (!fused_var.name().empty()) {
+        all_vars.push_back(fused_var);
+    }
+    if (!fused_rvar.name().empty()) {
+        all_vars.push_back(fused_rvar);
+    }
+    // Only reorder if there's more than one variables.
+    if (all_vars.size() > 1) {
+        func_or_stage.reorder(all_vars);
+        schedule_source << "    .reorder(";
+        for (int i = 0; i < (int)all_vars.size(); i++) {
+            schedule_source << all_vars[i].name();
+            if (i != (int)all_vars.size() - 1) {
+                schedule_source << ",";
+            }
+        }
+        schedule_source << ")\n";
+        if (is_pure_def) {
+            if (gpu_thread_dim > 0) {
+                std::vector<Var> reordered_vars = vars;
+                std::swap(reordered_vars[0], reordered_vars[gpu_thread_dim]);
+                reorder_storage(func_or_stage, reordered_vars, schedule_source);
+            }
+        }
+    }
+    
+    if (gpu_blocks.size() + r_gpu_blocks.size() > 0) { 
+        // Assign outer loops to GPU blocks
+        if (!fused_var.name().empty()) {
+            func_or_stage.gpu_blocks(fused_var);
+            schedule_source << "    .gpu_blocks(" << fused_var.name() << ")\n";
+        }
+        if (!fused_rvar.name().empty()) {
+            func_or_stage.atomic()
+                         .gpu_blocks(fused_rvar);
+            schedule_source << "    .atomic()\n";
+            schedule_source << "    .gpu_blocks(" << fused_rvar.name() << ")\n";
+        }
+        // Assign inner loops to GPU threads
+        if (!gpu_threads.name().empty()) {
+            func_or_stage.gpu_threads(gpu_threads);
+            schedule_source << "    .gpu_threads(" << gpu_threads.name() << ")\n";
+        }
+        if (!r_gpu_threads.name().empty()) {
+            func_or_stage.gpu_threads(r_gpu_threads);
+            schedule_source << "    .r_gpu_threads(" << r_gpu_threads.name() << ")\n";
+        }
+    } else {
+        // Not enough parallelism, use a single GPU thread
+        func_or_stage.gpu_single_thread();
+        schedule_source << "    .gpu_single_thread()\n";
+    }
+}
+
+
+template <typename FuncOrStage>
+void parallelize_vars_and_rvars_cpu(
+        const MachineParams &params,
+        FuncOrStage func_or_stage,
+        int natural_vector_size,
+        bool is_pure_def,
+        const std::vector<Var> &vars,
+        const std::vector<int> &var_bounds,
+        const std::vector<RVar> &rvars,
+        const std::vector<int> &rvar_bounds,
+        TailStrategy tail,
+        std::ostringstream &schedule_source) {
+    // Find the first variable that has bounds larger or equal than 16,
```

Comment:
comment says 16, but I don't see where 16 is actually referenced in code

Reply:
Fixed.